modules/ui/src/main/scala/lucuma/ui/forms/FormInputEV.scala

```diff
@@ -53,92 +55,135 @@ final case class FormInputEV[EV[_], A](
   transparent:     js.UndefOr[Boolean] = js.undefined,
   width:           js.UndefOr[SemanticWidth] = js.undefined,
   value:           EV[A],
-  format:          InputFormat[A] = InputFormat.id,
+  validate:        InputValidate[A] = InputValidate.id,
   modifiers:       Seq[TagMod] = Seq.empty,
-  onChange:        FormInputEV.ChangeCallback[A] =
-    (_: A) => Callback.empty, // callback for parents of this component
-  onBlur:          FormInputEV.ChangeCallback[A] = (_: A) => Callback.empty
+  onChange:        FormInputEV.ChangeCallback[A] = (_: A) => Callback.empty, // for extra actions
+  onValidChange:   FormInputEV.ChangeCallback[Boolean] = _ => Callback.empty,
+  onBlur:          FormInputEV.ChangeCallback[A] = (_: A) => Callback.empty // for extra actions
 )(implicit val ev: ExternalValue[EV])
     extends ReactProps[FormInputEV[Any, Any]](FormInputEV.component) {
-  def valGet: String = ev.get(value).foldMap(format.reverseGet)
-  def valSet(s: String): Callback =
-    format.getOption(s).map(ev.set(value)).getOrEmpty
-  val onBlurC: InputEV.ChangeCallback[String]   =
-    (s: String) => format.getOption(s).map(onBlur).getOrEmpty
-  val onChangeC: InputEV.ChangeCallback[String] =
-    (s: String) => format.getOption(s).map(onChange).getOrEmpty
+
+  def valGet: String = ev.get(value).foldMap(validate.reverseGet)
+
+  def valSet: InputEV.ChangeCallback[A] = ev.set(value)
+
+  def onBlurC(onError: NonEmptyList[String] => Callback): InputEV.ChangeCallback[String] =
+    (s: String) =>
+      validate
+        .getValidated(s)
+        .fold(onError, onBlur)
 
   def withMods(mods: TagMod*): FormInputEV[EV, A] = copy(modifiers = modifiers ++ mods)
 }
 
 object FormInputEV {
   type Props[EV[_], A]   = FormInputEV[EV, A]
   type ChangeCallback[A] = A => Callback
-  type Scope[EV[_], A]   = RenderScope[Props[EV, A], State, Unit]
 
   @Lenses
-  final case class State(curValue: String, prevValue: String)
-
-  def onTextChange[EV[_], A]($ : Scope[EV, A]): ReactEventFromInput => Callback =
-    (e: ReactEventFromInput) => {
-      // Capture the value outside setState, react reuses the events
-      val v = e.target.value
-      // First update the internal state, then call the outside listener
-      $.setStateL(State.curValue)(v) *>
-        // Next 2 might not be called if the InputFormat returns None
-        $.props.valSet(v) *>
-        $.props.onChangeC(v)
+  final case class State(curValue: String, prevValue: String, errors: Option[NonEmptyList[String]])
```

Comment:
It doesn't make much difference here but there is also `NonEmptyChain`. I'm using it on `lucuma-catalog`

Reply:
Yes, `NonEmptyChain` allows concatenating in constant time. We can switch to it, but It probably won't make much of a difference since here we are only processing validations for one control, unless there's a lot of validations for one control.